arangod/RocksDBEngine/RocksDBChecksumEnv.cpp

```diff
@@ -0,0 +1,264 @@
+  TRI_ASSERT(_context != nullptr);
+  unsigned char hash[EVP_MAX_MD_SIZE];
+  unsigned int lengthOfHash = 0;
+  if (EVP_DigestFinal_ex(_context, hash, &lengthOfHash) == 0) {
+    TRI_ASSERT(false);
+  }
+  _checksum = basics::StringUtils::encodeHex(
+      reinterpret_cast<char const*>(&hash[0]), lengthOfHash);
+}
+
+void ChecksumCalculator::updateIncrementalChecksum(char const* buffer,
+                                                   size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  updateEVPWithContent(buffer, n);
+}
+
+ChecksumCalculator::~ChecksumCalculator() {
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+  EVP_MD_CTX_free(_context);
+#else
+  EVP_MD_CTX_destroy(_context);
+#endif
+}
+
+void ChecksumCalculator::updateEVPWithContent(const char* buffer, size_t n) {
+  TRI_ASSERT(_context != nullptr);
+  if (EVP_DigestUpdate(_context, static_cast<void const*>(buffer), n) == 0) {
+    TRI_ASSERT(false);
+  }
+}
+
+bool ChecksumHelper::isFileNameSst(std::string const& fileName) {
+  return TRI_Basename(fileName).size() > 4 &&
+         (fileName.compare(fileName.size() - 4, 4, ".sst") == 0);
+}
+
+bool ChecksumHelper::writeShaFile(std::string const& fileName,
+                                  std::string const& checksum) {
+  TRI_ASSERT(isFileNameSst(fileName));
+
+  std::string shaFileName = buildShaFileNameFromSst(fileName, checksum);
+
+  LOG_TOPIC("80257", DEBUG, arangodb::Logger::ENGINES)
+      << "shaCalcFile: done " << fileName << " result: " << shaFileName;
+  auto res = TRI_WriteFile(shaFileName.c_str(), "", 0);
+  if (res == TRI_ERROR_NO_ERROR) {
+    {
+      MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+      _fileNamesToHashes.try_emplace(TRI_Basename(fileName), checksum);
+    }
+    return true;
+  }
+
+  LOG_TOPIC("8f7ef", WARN, arangodb::Logger::ENGINES)
+      << "shaCalcFile: writing file failed with " << res << " for "
+      << shaFileName;
+  return false;
+}
+
+void ChecksumHelper::checkMissingShaFiles() {
+  if (!_rootPath.empty()) {
+    std::vector<std::string> fileList = TRI_FilesDirectory(_rootPath.c_str());
+    std::sort(fileList.begin(), fileList.end());
+    std::string sstFileName;
+    for (auto it = fileList.begin(); it != fileList.end(); ++it) {
+      if (it->size() < 5) {
+        // filename is too short and does not matter
+        continue;
+      }
+      TRI_ASSERT(*it == TRI_Basename(*it));
+      std::string::size_type shaIndex = it->find(".sha.");
+      if (shaIndex != std::string::npos) {
+        sstFileName = it->substr(0, shaIndex) + ".sst";
+        TRI_ASSERT(sstFileName == TRI_Basename(sstFileName));
+        auto nextIt = it + 1;
+        if (nextIt != fileList.end() && nextIt->compare(sstFileName) == 0) {
+          TRI_ASSERT(it->size() >= shaIndex + 64);
+          std::string hash = it->substr(shaIndex + /*.sha.*/ 5, 64);
+          it = nextIt;
+          MUTEX_LOCKER(mutexLock, _calculatedHashesMutex);
+          _fileNamesToHashes.try_emplace(sstFileName, std::move(hash));
```

Comment:
AFAIK `sstFileName` it is not used afterwards, so it could be moved.
```suggestion
          _fileNamesToHashes.try_emplace(std::move(sstFileName), std::move(hash));
```